src/Servers/IIS/IIS/src/Core/IISHttpContext.cs

```diff
@@ -119,71 +119,77 @@ internal abstract partial class IISHttpContext : NativeRequestContext, IThreadPo
+                    Path = originalPath;
+                }
 
-            var cookedUrl = GetCookedUrl();
-            QueryString = cookedUrl.GetQueryString() ?? string.Empty;
+                var cookedUrl = GetCookedUrl();
+                QueryString = cookedUrl.GetQueryString() ?? string.Empty;
 
-            RequestHeaders = new RequestHeaders(this);
-            HttpResponseHeaders = new HeaderCollection();
-            ResponseHeaders = HttpResponseHeaders;
+                RequestHeaders = new RequestHeaders(this);
+                HttpResponseHeaders = new HeaderCollection();
+                ResponseHeaders = HttpResponseHeaders;
 
-            if (_options.ForwardWindowsAuthentication)
-            {
-                WindowsUser = GetWindowsPrincipal();
-                if (_options.AutomaticAuthentication)
+                if (_options.ForwardWindowsAuthentication)
                 {
-                    User = WindowsUser;
+                    WindowsUser = GetWindowsPrincipal();
+                    if (_options.AutomaticAuthentication)
+                    {
+                        User = WindowsUser;
+                    }
                 }
-            }
 
-            MaxRequestBodySize = _options.MaxRequestBodySize;
+                MaxRequestBodySize = _options.MaxRequestBodySize;
 
-            ResetFeatureCollection();
+                ResetFeatureCollection();
 
-            if (!_server.IsWebSocketAvailable(_pInProcessHandler))
-            {
-                _currentIHttpUpgradeFeature = null;
-            }
+                if (!_server.IsWebSocketAvailable(_pInProcessHandler))
+                {
+                    _currentIHttpUpgradeFeature = null;
+                }
 
-            _streams = new Streams(this);
+                _streams = new Streams(this);
 
-            (RequestBody, ResponseBody) = _streams.Start();
+                (RequestBody, ResponseBody) = _streams.Start();
 
-            var pipe = new Pipe(
-                new PipeOptions(
-                    _memoryPool,
-                    readerScheduler: PipeScheduler.ThreadPool,
-                    pauseWriterThreshold: PauseWriterThreshold,
-                    resumeWriterThreshold: ResumeWriterTheshold,
-                    minimumSegmentSize: MinAllocBufferSize));
-            _bodyOutput = new OutputProducer(pipe);
+                var pipe = new Pipe(
+                    new PipeOptions(
+                        _memoryPool,
+                        readerScheduler: PipeScheduler.ThreadPool,
+                        pauseWriterThreshold: PauseWriterThreshold,
+                        resumeWriterThreshold: ResumeWriterTheshold,
+                        minimumSegmentSize: MinAllocBufferSize));
+                _bodyOutput = new OutputProducer(pipe);
 
-            NativeMethods.HttpSetManagedContext(_pInProcessHandler, (IntPtr)_thisHandle);
+                NativeMethods.HttpSetManagedContext(_pInProcessHandler, (IntPtr)_thisHandle);
```

Comment:
Nit: Moving HttpSetManagedContext after the lock avoids possible (rare) contention.